src/evaluators/ForStatement.js

```diff
@@ -268,13 +280,152 @@ function ForBodyEvaluation(
+
+  if (functionInfo.usesReturn || functionInfo.usesArguments || functionInfo.usesGotoToLabel) {
+    // We do not have support for these yet
+    // TODO: issue number upon creation of issue
+    throw new FatalError("TODO: #? handle more loop bail out cases");
+  }
+  let args = [wrapperFunction];
+
+  if (functionInfo.usesThis) {
+    let thisRef = env.evaluate(t.thisExpression(), strictCode);
+    if (thisRef instanceof Value) {
+      Havoc.value(realm, thisRef);
+      args.push(thisRef);
+    }
+  }
+
+  Havoc.value(realm, wrapperFunction);
+
+  let wrapperValue = AbstractValue.createTemporalFromBuildFunction(
+    realm,
+    Value,
+    args,
+    ([func, thisExpr]) =>
+      functionInfo.usesThis
+        ? t.callExpression(t.memberExpression(func, t.identifier("call")), [thisExpr])
+        : t.callExpression(func, [])
+  );
+  invariant(wrapperValue instanceof AbstractValue);
+  return wrapperValue;
+}
+
+function tryToEvaluateForStatementOrLeaveAsAbstract(
+  ast: BabelNodeForStatement,
+  strictCode: boolean,
+  env: LexicalEnvironment,
+  realm: Realm,
+  labelSet: ?Array<string>
+): Value {
+  let effects;
+  let savedSuppressDiagnostics = realm.suppressDiagnostics;
+  try {
+    realm.suppressDiagnostics = true;
+    effects = realm.evaluateForEffects(
+      () => evaluateForStatement(ast, strictCode, env, realm, labelSet),
+      undefined,
+      "tryToEvaluateForStatementOrLeaveAsAbstract"
+    );
+  } catch (error) {
+    if (error instanceof FatalError) {
+      realm.suppressDiagnostics = savedSuppressDiagnostics;
+      return realm.evaluateWithPossibleThrowCompletion(
+        () => generateRuntimeForStatement(ast, strictCode, env, realm, labelSet),
+        TypesDomain.topVal,
+        ValuesDomain.topVal
+      );
+    } else {
+      throw error;
+    }
+  } finally {
+    realm.suppressDiagnostics = savedSuppressDiagnostics;
+  }
+  // Note that the effects of (non joining) abrupt branches are not included
+  // in effects, but are tracked separately inside completion.
+  realm.applyEffects(effects);
+  let completion = effects.result;
+  if (completion instanceof PossiblyNormalCompletion) {
+    // in this case one of the branches may complete abruptly, which means that
+    // not all control flow branches join into one flow at this point.
+    // Consequently we have to continue tracking changes until the point where
+    // all the branches come together into one.
+    completion = realm.composeWithSavedCompletion(completion);
+  }
+  // return or throw completion
+  if (completion instanceof AbruptCompletion) throw completion;
+  invariant(completion instanceof Value);
+  return completion;
+}
+
 // ECMA262 13.7.4.7
 export default function(
   ast: BabelNodeForStatement,
   strictCode: boolean,
   env: LexicalEnvironment,
   realm: Realm,
   labelSet: ?Array<string>
+): Value {
+  if (realm.isInPureScope()) {
+    return tryToEvaluateForStatementOrLeaveAsAbstract(ast, strictCode, env, realm, labelSet);
```

Comment:
Not a change request, just a thought: There's a growing set of bail-outs from just regular JavaScript code. It would be good to have some kind of logging or account for when these things happen, maybe combined with some configuration as to what kind of things are allowed. (We don't want to allow bailing out for loops or arbitrary functions in InstantRender, and I guess eventually you don't want that for the React Compiler either.) Also possibly just to give de-optimization hints to users.